File: python/ferro_ta/analysis/attribution.py
"""
ferro_ta.attribution — Performance attribution and trade analysis.
=================================================================

Compute trade-level statistics and attribute equity-curve performance to
individual signals or time periods.  Designed to work with the output of
``ferro_ta.backtest.backtest()``.

Functions
---------
trade_stats(pnl, hold_bars)
    Compute win rate, avg win/loss, profit factor, and avg hold duration.

from_backtest(result)
    Extract the trade list (PnL per trade, hold duration) from a
    :class:`~ferro_ta.backtest.BacktestResult`.

attribution_by_month(bar_returns, timestamps)
    Attribute per-bar returns to calendar months.

attribution_by_signal(bar_returns, signal_labels)
    Attribute per-bar returns to signal labels.

TradeStats
    Named-tuple-style result container returned by ``trade_stats``.

Rust backend
------------
    ferro_ta._ferro_ta.trade_stats
    ferro_ta._ferro_ta.monthly_contribution
    ferro_ta._ferro_ta.signal_attribution
"""

from __future__ import annotations

from typing import Any, Optional

import numpy as np
from numpy.typing import ArrayLike, NDArray

from ferro_ta._ferro_ta import (
    extract_trades as _rust_extract_trades,
)
from ferro_ta._ferro_ta import (
    monthly_contribution as _rust_monthly_contribution,
)
from ferro_ta._ferro_ta import (
    signal_attribution as _rust_signal_attribution,
)
from ferro_ta._ferro_ta import (
    trade_stats as _rust_trade_stats,
)
from ferro_ta._utils import _to_f64

__all__ = [
    "TradeStats",
    "trade_stats",
    "from_backtest",
    "attribution_by_month",
    "attribution_by_signal",
]


# ---------------------------------------------------------------------------
# TradeStats container
# ---------------------------------------------------------------------------


class TradeStats:
    """Container for trade-level statistics.

    Attributes
    ----------
    win_rate      : float — fraction of trades with PnL > 0
    avg_win       : float — mean PnL of winning trades (0 if none)
    avg_loss      : float — mean PnL of losing trades (negative; 0 if none)
    profit_factor : float — gross profit / |gross loss|  (inf if no losses)
    avg_hold_bars : float — mean hold duration in bars
    n_trades      : int   — total number of trades
    """

    __slots__ = (
        "win_rate",
        "avg_win",
        "avg_loss",
        "profit_factor",
        "avg_hold_bars",
        "n_trades",
    )

    def __init__(
        self,
        win_rate: float,
        avg_win: float,
        avg_loss: float,
        profit_factor: float,
        avg_hold_bars: float,
        n_trades: int,
    ) -> None:
        self.win_rate = win_rate
        self.avg_win = avg_win
        self.avg_loss = avg_loss
        self.profit_factor = profit_factor
        self.avg_hold_bars = avg_hold_bars
        self.n_trades = n_trades

    def __repr__(self) -> str:
        return (
            f"TradeStats(n_trades={self.n_trades}, "
            f"win_rate={self.win_rate:.2%}, "
            f"profit_factor={self.profit_factor:.2f}, "
            f"avg_hold={self.avg_hold_bars:.1f} bars)"
        )

    def to_dict(self) -> dict[str, Any]:
        """Return stats as a plain dict."""
        return {
            "n_trades": self.n_trades,
            "win_rate": self.win_rate,
            "avg_win": self.avg_win,
            "avg_loss": self.avg_loss,
            "profit_factor": self.profit_factor,
            "avg_hold_bars": self.avg_hold_bars,
        }


# ---------------------------------------------------------------------------
# trade_stats
# ---------------------------------------------------------------------------


def trade_stats(
    pnl: ArrayLike,
    hold_bars: Optional[ArrayLike] = None,
) -> TradeStats:
    """Compute trade-level performance statistics.

    Parameters
    ----------
    pnl       : array-like — per-trade PnL (positive = win, negative = loss)
    hold_bars : array-like, optional — hold duration in bars for each trade.
        If ``None``, defaults to an array of ones (hold duration unknown).

    Returns
    -------
    :class:`TradeStats`

    Examples
    --------
    >>> import numpy as np
    >>> from ferro_ta.analysis.attribution import trade_stats
    >>> pnl = np.array([10.0, -5.0, 8.0, -3.0, 15.0, -2.0])
    >>> hold = np.array([5.0, 3.0, 7.0, 2.0, 10.0, 1.0])
    >>> ts = trade_stats(pnl, hold)
    >>> print(ts)
    TradeStats(n_trades=6, win_rate=50.00%, profit_factor=...)
    """
    p = _to_f64(pnl)
    n = len(p)
    if n == 0:
        raise ValueError("pnl must be non-empty")
    if hold_bars is None:
        h = np.ones(n, dtype=np.float64)
    else:
        h = _to_f64(hold_bars)

    win_rate, avg_win, avg_loss, profit_factor, avg_hold = _rust_trade_stats(p, h)
    return TradeStats(
        win_rate=win_rate,
        avg_win=avg_win,
        avg_loss=avg_loss,
        profit_factor=profit_factor,
        avg_hold_bars=avg_hold,
        n_trades=n,
    )


# ---------------------------------------------------------------------------
# from_backtest
# ---------------------------------------------------------------------------


def from_backtest(result: Any) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
    """Extract per-trade PnL and hold durations from a BacktestResult.

    Scans the ``positions`` and ``strategy_returns`` arrays of *result* to
    find trade entries and exits, then computes per-trade PnL and duration.

    Parameters
    ----------
    result : :class:`~ferro_ta.backtest.BacktestResult`

    Returns
    -------
    tuple ``(pnl, hold_bars)`` — 1-D float64 arrays of length n_trades.

    Notes
    -----
    A "trade" is defined as a continuous run of non-zero position.  PnL is
    the sum of ``strategy_returns`` during that period.  Hold duration is
    the number of bars in the run.
    """
    pos = np.asarray(result.positions, dtype=np.float64)
    ret = np.asarray(result.strategy_returns, dtype=np.float64)
    pnl, hold = _rust_extract_trades(pos, ret)
    return (
        np.asarray(pnl, dtype=np.float64),
        np.asarray(hold, dtype=np.float64),
    )


# ---------------------------------------------------------------------------
# attribution_by_month
# ---------------------------------------------------------------------------


def attribution_by_month(
    bar_returns: ArrayLike,
    timestamps: Optional[ArrayLike] = None,
) -> dict[str, float]:
    """Attribute per-bar returns to calendar months.

    Parameters
    ----------
    bar_returns : array-like — per-bar strategy returns
    timestamps  : array-like of int64, optional — UTC timestamps in
        nanoseconds (e.g. ``pandas.DatetimeIndex.astype('int64')``).
        If ``None``, bars are grouped into calendar-agnostic monthly buckets
        of 21 bars (approximate trading month).

    Returns
    -------
    dict mapping month label (str ``'YYYY-MM'`` or ``'period_N'``) to
    total return for that month.

    Examples
    --------
    >>> import numpy as np
    >>> from ferro_ta.analysis.attribution import attribution_by_month
    >>> rng = np.random.default_rng(0)
    >>> ret = rng.normal(0, 0.01, 252)
    >>> contrib = attribution_by_month(ret)
    >>> list(contrib.keys())[:3]
    ['period_0', 'period_1', 'period_2']
    """
    ret = _to_f64(bar_returns)
    n = len(ret)

    if timestamps is not None:
        # Convert ns timestamps → month index
        ts = np.asarray(timestamps, dtype=np.int64)
        # Month = year*12 + month_of_year (0-based)
        # ns → seconds → datetime calculation (fast path without pandas)
        try:
            import pandas as pd

            dti = pd.to_datetime(ts, unit="ns", utc=True)
            month_idx = (dti.year * 12 + dti.month - 1).astype(np.int64)  # type: ignore[union-attr]
            offset = int(month_idx[0])
            month_idx = (month_idx - offset).values.astype(np.int64)
        except ImportError:
            # Fallback: 21-bar buckets
            month_idx = np.arange(n, dtype=np.int64) // 21
    else:
        month_idx = np.arange(n, dtype=np.int64) // 21

    months_arr, contrib_arr = _rust_monthly_contribution(ret, month_idx)
    months = np.asarray(months_arr, dtype=np.int64)
    contribs = np.asarray(contrib_arr, dtype=np.float64)

    if timestamps is not None:
        try:
            import pandas as pd

            ts = np.asarray(timestamps, dtype=np.int64)
            dti = pd.to_datetime(ts, unit="ns", utc=True)
            month_idx_full = (dti.year * 12 + dti.month - 1).astype(np.int64).values  # type: ignore[union-attr]
            offset = int(month_idx_full[0])
            labels = {}
            for m, c in zip(months, contribs):
                abs_month = int(m) + offset
                year = abs_month // 12
                month_of_year = abs_month % 12 + 1
                labels[f"{year:04d}-{month_of_year:02d}"] = float(c)
            return labels
        except ImportError:
            pass

    return {f"period_{int(m)}": float(c) for m, c in zip(months, contribs)}


# ---------------------------------------------------------------------------
# attribution_by_signal
# ---------------------------------------------------------------------------


def attribution_by_signal(
    bar_returns: ArrayLike,
    signal_labels: ArrayLike,
) -> dict[str, float]:
    """Attribute per-bar returns to signal labels.

    Parameters
    ----------
    bar_returns   : array-like — per-bar strategy returns
    signal_labels : array-like of int — signal label per bar.
        Use ``-1`` for flat (no position) bars.

    Returns
    -------
    dict mapping signal label (str) to total attributed return.

    Examples
    --------
    >>> import numpy as np
    >>> from ferro_ta.analysis.attribution import attribution_by_signal
    >>> rng = np.random.default_rng(0)
    >>> ret = rng.normal(0, 0.01, 100)
    >>> labels = np.where(np.arange(100) < 50, 0, 1)  # signal 0 or signal 1
    >>> contrib = attribution_by_signal(ret, labels)
    >>> sorted(contrib.keys())
    ['signal_0', 'signal_1']
    """
    ret = _to_f64(bar_returns)
    lbl = np.asarray(signal_labels, dtype=np.int64)
    labels_arr, contrib_arr = _rust_signal_attribution(ret, lbl)
    labels = np.asarray(labels_arr, dtype=np.int64)
    contribs = np.asarray(contrib_arr, dtype=np.float64)
    return {f"signal_{int(lbl)}": float(c) for lbl, c in zip(labels, contribs)}

File: crates/ferro_ta_core/src/attribution.rs
//! Performance attribution and trade analysis — pure Rust, no PyO3.
//!
//! Functions
//! ---------
//! - `trade_stats`          — win rate, avg win/loss, profit factor, avg hold
//! - `monthly_contribution` — group bar returns by month index and sum
//! - `signal_attribution`   — group bar returns by signal label and sum
//! - `extract_trades`       — extract trade pnl and hold durations from positions

use std::collections::HashMap;

// ---------------------------------------------------------------------------
// trade_stats
// ---------------------------------------------------------------------------

/// Compute trade-level statistics from trade PnL and hold durations.
///
/// Returns `(win_rate, avg_win, avg_loss, profit_factor, avg_hold_bars)`.
///
/// - **win_rate**      : fraction of trades with PnL > 0
/// - **avg_win**       : mean PnL of winning trades (0 if none)
/// - **avg_loss**      : mean PnL of losing trades (negative; 0 if none)
/// - **profit_factor** : gross profit / |gross loss| (inf if no losses)
/// - **avg_hold_bars** : mean hold duration across all trades
///
/// # Panics
/// Panics if `pnl` is empty or `pnl.len() != hold_bars.len()`.
pub fn trade_stats(pnl: &[f64], hold_bars: &[f64]) -> (f64, f64, f64, f64, f64) {
    let n = pnl.len();
    assert!(n > 0, "pnl must be non-empty");
    assert_eq!(
        n,
        hold_bars.len(),
        "pnl and hold_bars must have equal length"
    );

    let mut wins: Vec<f64> = Vec::new();
    let mut losses: Vec<f64> = Vec::new();
    for &v in pnl.iter() {
        if v > 0.0 {
            wins.push(v);
        } else if v < 0.0 {
            losses.push(v);
        }
    }

    let win_rate = wins.len() as f64 / n as f64;
    let avg_win = if wins.is_empty() {
        0.0
    } else {
        wins.iter().sum::<f64>() / wins.len() as f64
    };
    let avg_loss = if losses.is_empty() {
        0.0
    } else {
        losses.iter().sum::<f64>() / losses.len() as f64
    };

    let gross_profit: f64 = wins.iter().sum();
    let gross_loss: f64 = losses.iter().map(|v| v.abs()).sum();
    let profit_factor = if gross_loss == 0.0 {
        f64::INFINITY
    } else {
        gross_profit / gross_loss
    };

    let avg_hold = hold_bars.iter().sum::<f64>() / n as f64;

    (win_rate, avg_win, avg_loss, profit_factor, avg_hold)
}

// ---------------------------------------------------------------------------
// monthly_contribution
// ---------------------------------------------------------------------------

/// Group per-bar returns by month index and sum each month's contribution.
///
/// Returns `(months, contributions)` where `months` is sorted unique month
/// indices and `contributions` is the corresponding total return per month.
/// NaN returns are skipped.
///
/// # Panics
/// Panics if `bar_returns.len() != month_index.len()`.
pub fn monthly_contribution(bar_returns: &[f64], month_index: &[i64]) -> (Vec<i64>, Vec<f64>) {
    let n = bar_returns.len();
    assert_eq!(
        n,
        month_index.len(),
        "bar_returns and month_index must have equal length"
    );

    let mut map: HashMap<i64, f64> = HashMap::new();
    for i in 0..n {
        if !bar_returns[i].is_nan() {
            *map.entry(month_index[i]).or_insert(0.0) += bar_returns[i];
        }
    }

    let mut months: Vec<i64> = map.keys().copied().collect();
    months.sort_unstable();
    let contributions: Vec<f64> = months.iter().map(|m| map[m]).collect();

    (months, contributions)
}

// ---------------------------------------------------------------------------
// signal_attribution
// ---------------------------------------------------------------------------

/// Attribute per-bar returns to each signal label.
///
/// Returns `(labels, contributions)` where `labels` is sorted unique signal
/// labels and `contributions` is the corresponding total return per label.
/// NaN returns are skipped.
///
/// # Panics
/// Panics if `bar_returns.len() != signal_labels.len()`.
pub fn signal_attribution(bar_returns: &[f64], signal_labels: &[i64]) -> (Vec<i64>, Vec<f64>) {
    let n = bar_returns.len();
    assert_eq!(
        n,
        signal_labels.len(),
        "bar_returns and signal_labels must have equal length"
    );

    let mut map: HashMap<i64, f64> = HashMap::new();
    for i in 0..n {
        if !bar_returns[i].is_nan() {
            *map.entry(signal_labels[i]).or_insert(0.0) += bar_returns[i];
        }
    }

    let mut labels: Vec<i64> = map.keys().copied().collect();
    labels.sort_unstable();
    let contributions: Vec<f64> = labels.iter().map(|l| map[l]).collect();

    (labels, contributions)
}

// ---------------------------------------------------------------------------
// extract_trades
// ---------------------------------------------------------------------------

/// Extract trade-level PnL and hold durations from positions and strategy returns.
///
/// A trade is a maximal contiguous run of non-zero position values with the
/// same sign/magnitude. Returns `(pnl, hold_durations)`.
///
/// # Panics
/// Panics if `positions.len() != strategy_returns.len()`.
pub fn extract_trades(positions: &[f64], strategy_returns: &[f64]) -> (Vec<f64>, Vec<f64>) {
    let n = positions.len();
    assert_eq!(
        n,
        strategy_returns.len(),
        "positions and strategy_returns must have equal length"
    );

    let mut pnl = Vec::<f64>::new();
    let mut hold = Vec::<f64>::new();

    let mut i = 0usize;
    while i < n {
        if positions[i] == 0.0 {
            i += 1;
            continue;
        }
        let mut j = i + 1;
        while j < n && positions[j] == positions[i] {
            j += 1;
        }
        let mut trade_pnl = 0.0_f64;
        for v in strategy_returns.iter().take(j).skip(i) {
            trade_pnl += *v;
        }
        pnl.push(trade_pnl);
        hold.push((j - i) as f64);
        i = j;
    }

    (pnl, hold)
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    // -- trade_stats ---------------------------------------------------------

    #[test]
    fn test_trade_stats_basic() {
        let pnl = [100.0, -50.0, 200.0, -30.0, 150.0];
        let hold = [5.0, 3.0, 7.0, 2.0, 6.0];
        let (wr, aw, al, pf, ah) = trade_stats(&pnl, &hold);

        // 3 wins out of 5
        assert!((wr - 0.6).abs() < 1e-10);
        // avg win = (100+200+150)/3
        assert!((aw - 150.0).abs() < 1e-10);
        // avg loss = (-50 + -30)/2 = -40
        assert!((al - (-40.0)).abs() < 1e-10);
        // profit_factor = 450 / 80
        assert!((pf - 5.625).abs() < 1e-10);
        // avg hold = (5+3+7+2+6)/5 = 4.6
        assert!((ah - 4.6).abs() < 1e-10);
    }

    #[test]
    fn test_trade_stats_all_wins() {
        let pnl = [10.0, 20.0];
        let hold = [1.0, 2.0];
        let (wr, _aw, al, pf, _ah) = trade_stats(&pnl, &hold);
        assert!((wr - 1.0).abs() < 1e-10);
        assert!((al - 0.0).abs() < 1e-10);
        assert!(pf.is_infinite());
    }

    #[test]
    fn test_trade_stats_all_losses() {
        let pnl = [-10.0, -20.0];
        let hold = [1.0, 2.0];
        let (wr, aw, _al, pf, _ah) = trade_stats(&pnl, &hold);
        assert!((wr - 0.0).abs() < 1e-10);
        assert!((aw - 0.0).abs() < 1e-10);
        assert!((pf - 0.0).abs() < 1e-10);
    }

    #[test]
    #[should_panic(expected = "pnl must be non-empty")]
    fn test_trade_stats_empty() {
        trade_stats(&[], &[]);
    }

    // -- monthly_contribution ------------------------------------------------

    #[test]
    fn test_monthly_contribution_basic() {
        let returns = [0.01, 0.02, -0.01, 0.03, -0.02];
        let months = [0, 0, 1, 1, 2];
        let (m, c) = monthly_contribution(&returns, &months);
        assert_eq!(m, vec![0, 1, 2]);
        assert!((c[0] - 0.03).abs() < 1e-10);
        assert!((c[1] - 0.02).abs() < 1e-10);
        assert!((c[2] - (-0.02)).abs() < 1e-10);
    }

    #[test]
    fn test_monthly_contribution_nan_skipped() {
        let returns = [0.01, f64::NAN, 0.03];
        let months = [0, 0, 1];
        let (m, c) = monthly_contribution(&returns, &months);
        assert_eq!(m, vec![0, 1]);
        assert!((c[0] - 0.01).abs() < 1e-10);
        assert!((c[1] - 0.03).abs() < 1e-10);
    }

    #[test]
    fn test_monthly_contribution_empty() {
        let (m, c) = monthly_contribution(&[], &[]);
        assert!(m.is_empty());
        assert!(c.is_empty());
    }

    // -- signal_attribution --------------------------------------------------

    #[test]
    fn test_signal_attribution_basic() {
        let returns = [0.05, -0.02, 0.03, 0.01];
        let labels = [1, -1, 2, 1];
        let (l, c) = signal_attribution(&returns, &labels);
        assert_eq!(l, vec![-1, 1, 2]);
        assert!((c[0] - (-0.02)).abs() < 1e-10);
        assert!((c[1] - 0.06).abs() < 1e-10); // 0.05 + 0.01
        assert!((c[2] - 0.03).abs() < 1e-10);
    }

    #[test]
    fn test_signal_attribution_nan_skipped() {
        let returns = [0.05, f64::NAN];
        let labels = [1, 2];
        let (l, c) = signal_attribution(&returns, &labels);
        assert_eq!(l, vec![1]);
        assert!((c[0] - 0.05).abs() < 1e-10);
    }

    // -- extract_trades ------------------------------------------------------

    #[test]
    fn test_extract_trades_basic() {
        // positions: flat, long, long, flat, short, short
        let positions = [0.0, 1.0, 1.0, 0.0, -1.0, -1.0];
        let strat_ret = [0.0, 0.01, 0.02, 0.0, -0.01, 0.03];
        let (pnl, hold) = extract_trades(&positions, &strat_ret);
        assert_eq!(pnl.len(), 2);
        assert_eq!(hold.len(), 2);
        // First trade: bars 1..3 => 0.01 + 0.02 = 0.03
        assert!((pnl[0] - 0.03).abs() < 1e-10);
        assert!((hold[0] - 2.0).abs() < 1e-10);
        // Second trade: bars 4..6 => -0.01 + 0.03 = 0.02
        assert!((pnl[1] - 0.02).abs() < 1e-10);
        assert!((hold[1] - 2.0).abs() < 1e-10);
    }

    #[test]
    fn test_extract_trades_all_flat() {
        let positions = [0.0, 0.0, 0.0];
        let strat_ret = [0.01, 0.02, 0.03];
        let (pnl, hold) = extract_trades(&positions, &strat_ret);
        assert!(pnl.is_empty());
        assert!(hold.is_empty());
    }

    #[test]
    fn test_extract_trades_empty() {
        let (pnl, hold) = extract_trades(&[], &[]);
        assert!(pnl.is_empty());
        assert!(hold.is_empty());
    }

    #[test]
    fn test_extract_trades_single_bar_trade() {
        let positions = [0.0, 1.0, 0.0];
        let strat_ret = [0.0, 0.05, 0.0];
        let (pnl, hold) = extract_trades(&positions, &strat_ret);
        assert_eq!(pnl.len(), 1);
        assert!((pnl[0] - 0.05).abs() < 1e-10);
        assert!((hold[0] - 1.0).abs() < 1e-10);
    }
}
